## Content

- **Universal Functions (ufunc) on 2D**
    - Aggregate Function/ Reduction functions - `sum()`, `mean()`, `min()`, `max()`
    
    - Axis argument

    - Logical Operations
    
    - Sorting function - `sort()`, `argsort()`

    
       
       
- **Use Case: Fitness Data analysis**
    - Loading data set and EDA using numpy
    - `np.argmax()`

- **Reshape with -ve index**



- **Matrix Multiplication**
    - `matmul()`, `@`, `dot()`

    




## Universal Functions (`ufunc`) on 2D & Axis 

### Aggregate Functions/ Reduction functions

We saw how aggregate functions work on 1D array in last class



In [1]:
import numpy as np

In [2]:
arr = np.arange(3)
arr

array([0, 1, 2])

In [3]:
arr.sum()

3

#### Let's apply Aggregate functions on 2D array

### `np.sum()`

In [4]:
a = np.arange(12).reshape(3, 4)
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [5]:
np.sum(a)  # sums all the values present in array

66

#### What if we want to do the elements row-wise or column-wise?

- By **setting `axis` parameter**

#### What will `np.sum(a, axis=0)` do?

- **`np.sum(a, axis=0)` adds together values in DIFFERENT rows**
- **`axis = 0` ---> Changes will happen along the vertical axis**
- Summing of values happen **in the vertical direction**
- Rows collapse/merge when we do `axis=0`


In [6]:
np.sum(a, axis=0)

array([12, 15, 18, 21])

#### Now, What if we specify `axis=1`?

- **`np.sum(a, axis=1)` adds together values in DIFFERENT columns** 
- **`axis = 1` ---> Changes will happen along the horizontal axis**
- Summing of values happen **in the horizontal direction**
- Columns collapse/merge when we do `axis=1`

In [7]:
np.sum(a, axis=1)

array([ 6, 22, 38])

***

#### Now, What if we want to find the average value or median value of all the elements in an array?

In [8]:
np.mean(a) # no need to give any axis

5.5

#### What if we want to find the mean of elements in each row or in each column?

- We can do **same thing with `axis` parameter** like we did for `np.sum()` function


#### Question: Now you tell What will `np.mean(a, axis=0)` give?

- It will give **mean of values in DIFFERENT rows**
- **`axis = 0` ---> Changes will happen along the vertical axis**
- Mean of values will be calculated **in the vertical direction**
- Rows collapse/merge when we do `axis=0`

In [9]:
np.mean(a, axis=0)

array([4., 5., 6., 7.])

#### How can we get mean of elements in each column?

- **`np.mean(a, axis=1)` will give mean of values in DIFFERENT columns** 
- **`axis = 1` ---> Changes will happen along the horizontal axis**
- Mean of values will be calculated **in the horizontal direction**
- Columns collapse/merge when we do `axis=1`

In [10]:
np.mean(a, axis=1)

array([1.5, 5.5, 9.5])

#### Now, we want to find the minimum value in the array

**`np.min()` function can help us with this**

In [11]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [12]:
np.min(a)

0

#### What if we want to find row wise minimum value?

### Use `axis` argument!!

In [13]:
np.min(a, axis = 1 )

array([0, 4, 8])

#### We can also find max elements in an array.

**`np.max()`** function will give us *maximum value in the array*

We can also use `axis` argument to find row wise/ column wise max.

In [14]:
a

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [15]:
np.max(a) # maximum value

11

In [16]:
np.max(a, axis = 0) # column wise max

array([ 8,  9, 10, 11])

### Logical Operations

#### Now, What if we want to check whether "any" element of array follows a specific condition?

#### Let's say we have 2 arrays:

In [17]:
a = np.array([1,2,3,4])
b = np.array([4,3,2,1])
a, b

(array([1, 2, 3, 4]), array([4, 3, 2, 1]))

#### Let's say we want to find out if any of the elements in array `a` is smaller than any of the corresponding elements in array `b`


#### `np.any()` can become handy here as well


- `any()` returns `True` if **any of the corresponding elements** in the argument arrays follow the **provided condition**.


In [18]:
a = np.array([1,2,3,4])
b = np.array([4,3,2,1])
np.any(a<b) # Atleast 1 element in a < corresponding element in b

True

#### Let's try the same condition with different arrays:

In [24]:
a = np.array([4,5,6,7])
b = np.array([4,3,2,7])
np.any(a<b) # All elements in a >= corresponding elements in b

False

- In this case, **NONE of the elements in `a` were smaller than their corresponding elements in `b`**

- So, `np.any(a<b)` returned `False`

***

#### What if we want to check whether "all" the elements in our array are non-zero or follow the specified condition?

`np.all()`


#### Now, What if we want to check whether "all" the elements in our array follow a specific condition?




#### Let's say we want to find out if all the elements in array `a` are smaller than all the corresponding elements in array `b`

Again, Let's say we have 2 arrays:


In [29]:
a = np.array([1,2,3,4])
b = np.array([4,3,6,4])
a, b

(array([1, 2, 3, 4]), array([4, 3, 6, 4]))

In [30]:
np.all(a<b) # Not all elements in a < corresponding elements in b

False

#### Let's try it with different arrays

In [31]:
a = np.array([1,0,0,0])
b = np.array([4,3,2,1])
np.all(a<b) # All elements in a < corresponding elements in b

True

- In this case, **ALL the elements in `a` were smaller than their corresponding elements in `b`**

- So, `np.all(a<b)` returned `True`

#### Multiple conditions for `.all()` function

In [32]:
a = np.array([1, 2, 3, 2])
b = np.array([2, 2, 3, 2])
c = np.array([6, 4, 4, 5])
((a <= b) & (b <= c)).all()

True

#### What if we want to update an array based on condition ? 

Suppose you are given an array of integers and you want to update it based on following condition:
- if element is > 0, change it to +1
- if element < 0, change it to -1.

#### How will you do it ? 


In [41]:
arr = np.array([-3,4,27,34,-2, 0, -45,-11,4, 0 ])
arr

array([ -3,   4,  27,  34,  -2,   0, -45, -11,   4,   0])

You can use masking to update the array (as discussed in last class)

In [42]:
arr[arr > 0]  = 1
arr [arr < 0] = -1

In [43]:
arr

array([-1,  1,  1,  1, -1,  0, -1, -1,  1,  0])

There is a numpy function which can help us with it.

#### np.where()

Function signature: 
`np.where(condition, [x, y])`

This functions returns an ndarray whose elements are chosen from x or y depending on condition.



In [46]:
arr = np.array([-3,4,27,34,-2, 0, -45,-11,4, 0 ])


In [49]:
np.where(arr > 0, 'Hi', 'Bye')

array(['Bye', 'Hi', 'Hi', 'Hi', 'Bye', 'Bye', 'Bye', 'Bye', 'Hi', 'Bye'],
      dtype='<U3')

In [50]:
arr

array([ -3,   4,  27,  34,  -2,   0, -45, -11,   4,   0])

Notice that it didn't change the original array.

### Sorting Arrays

- We can also sort the elements of an array along a given specified axis 


- Default axis is the last axis of the array.



#### `np.sort()`



In [51]:
a = np.array([2,30,41,7,17,52])
a

array([ 2, 30, 41,  7, 17, 52])

In [52]:
np.sort(a)

array([ 2,  7, 17, 30, 41, 52])

In [53]:
a

array([ 2, 30, 41,  7, 17, 52])

Let's work with 2D array

In [54]:
a = np.arange(9,0,-1).reshape(3,3)
a

array([[9, 8, 7],
       [6, 5, 4],
       [3, 2, 1]])

#### Question: What will be the result when we sort using axis = 0 ?

In [55]:
np.sort(a, axis = 0)

array([[3, 2, 1],
       [6, 5, 4],
       [9, 8, 7]])

Recall that when axis =0
- change will happen along vertical axis.

Hence, it will sort out row wise.

In [ ]:
a

- Original array is still the same. It hasn't changed

#### `np.argsort()`

- Returns the **indices** that would sort an array.

- Performs an indirect sort along the given axis. 

- It returns **an array of indices of the same shape as a that index data along the given axis in sorted order**.

In [71]:
a = np.array([2,30,41,7,17,52])
a

array([ 2, 30, 41,  7, 17, 52])

In [72]:
np.argsort(a)

array([0, 3, 4, 1, 2, 5])

#### As you can see:

- The orginal indices of elements are in same order as the orginal elements would be in sorted order

## Use Case: Fitness data analysis



In [73]:
# !wget https://drive.google.com/uc?id=1vk1Pu0djiYcrdc85yUXZ_Rqq2oZNcohd -O ../fitbit.txt

### Let's load the data we saw earlier. For this we will use `.loadtxt() function`

In [74]:
data = np.loadtxt('../fit.txt', dtype='str')

We provide file name along with the dtype of data we want to load in 

In [75]:
data[:5]

array([['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
       ['07-10-2017', '6041', 'Sad', '197', '8', 'Inactive'],
       ['08-10-2017', '25', 'Sad', '0', '5', 'Inactive'],
       ['09-10-2017', '5461', 'Sad', '174', '4', 'Inactive'],
       ['10-10-2017', '6915', 'Neutral', '223', '5', 'Active']],
      dtype='<U10')

What's the shape of the data? 

In [76]:
data.shape

(96, 6)

There are 96 records and each record has 6 features.
These features are:
- Date
- Step count
- Mood
- Calories Burned
- Hours of sleep
- activity status

#### Notice that above array is a homogenous containing all the data as strings

In order to work with strings, categorical data and numerical data, we will have save every feature seperately

#### How will we extract features in seperate variables?

We can get some idea on how data is saved.

Lets see whats the first element of `data`

In [77]:
data[0]

array(['06-10-2017', '5464', 'Neutral', '181', '5', 'Inactive'],
      dtype='<U10')

Hm, this extracts a row not a column

Think about it.

#### Whats the way to change columns to rows and rows to columns?

Transpose

In [82]:
data[:,0] == data.T[0]

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True])

In [79]:
data.T[0]

array(['06-10-2017', '07-10-2017', '08-10-2017', '09-10-2017',
       '10-10-2017', '11-10-2017', '12-10-2017', '13-10-2017',
       '14-10-2017', '15-10-2017', '16-10-2017', '17-10-2017',
       '18-10-2017', '19-10-2017', '20-10-2017', '21-10-2017',
       '22-10-2017', '23-10-2017', '24-10-2017', '25-10-2017',
       '26-10-2017', '27-10-2017', '28-10-2017', '29-10-2017',
       '30-10-2017', '31-10-2017', '01-11-2017', '02-11-2017',
       '03-11-2017', '04-11-2017', '05-11-2017', '06-11-2017',
       '07-11-2017', '08-11-2017', '09-11-2017', '10-11-2017',
       '11-11-2017', '12-11-2017', '13-11-2017', '14-11-2017',
       '15-11-2017', '16-11-2017', '17-11-2017', '18-11-2017',
       '19-11-2017', '20-11-2017', '21-11-2017', '22-11-2017',
       '23-11-2017', '24-11-2017', '25-11-2017', '26-11-2017',
       '27-11-2017', '28-11-2017', '29-11-2017', '30-11-2017',
       '01-12-2017', '02-12-2017', '03-12-2017', '04-12-2017',
       '05-12-2017', '06-12-2017', '07-12-2017', '08-12

Great, we could extract first column

#### Lets extract all the columns and save them in seperate variables

In [83]:
date, step_count, mood, calories_burned, hours_of_sleep, activity_status = data.T

In [84]:
step_count

array(['5464', '6041', '25', '5461', '6915', '4545', '4340', '1230', '61',
       '1258', '3148', '4687', '4732', '3519', '1580', '2822', '181',
       '3158', '4383', '3881', '4037', '202', '292', '330', '2209',
       '4550', '4435', '4779', '1831', '2255', '539', '5464', '6041',
       '4068', '4683', '4033', '6314', '614', '3149', '4005', '4880',
       '4136', '705', '570', '269', '4275', '5999', '4421', '6930',
       '5195', '546', '493', '995', '1163', '6676', '3608', '774', '1421',
       '4064', '2725', '5934', '1867', '3721', '2374', '2909', '1648',
       '799', '7102', '3941', '7422', '437', '1231', '1696', '4921',
       '221', '6500', '3575', '4061', '651', '753', '518', '5537', '4108',
       '5376', '3066', '177', '36', '299', '1447', '2599', '702', '133',
       '153', '500', '2127', '2203'], dtype='<U10')

In [85]:
step_count.dtype

dtype('<U10')

Notice the data type of step_count and other variables. It's a string type where **U** means Unicode String. and 10 means 10 bytes. 

#### Why? Because Numpy type-casted all the data to strings.

#### Let's convert the data types of these variables

**Step Count**

In [86]:
step_count = np.array(step_count, dtype = 'int')
step_count.dtype

dtype('int64')

In [87]:
step_count

array([5464, 6041,   25, 5461, 6915, 4545, 4340, 1230,   61, 1258, 3148,
       4687, 4732, 3519, 1580, 2822,  181, 3158, 4383, 3881, 4037,  202,
        292,  330, 2209, 4550, 4435, 4779, 1831, 2255,  539, 5464, 6041,
       4068, 4683, 4033, 6314,  614, 3149, 4005, 4880, 4136,  705,  570,
        269, 4275, 5999, 4421, 6930, 5195,  546,  493,  995, 1163, 6676,
       3608,  774, 1421, 4064, 2725, 5934, 1867, 3721, 2374, 2909, 1648,
        799, 7102, 3941, 7422,  437, 1231, 1696, 4921,  221, 6500, 3575,
       4061,  651,  753,  518, 5537, 4108, 5376, 3066,  177,   36,  299,
       1447, 2599,  702,  133,  153,  500, 2127, 2203])

**Calories Burned**

In [91]:
calories_burned = np.array(calories_burned, dtype = 'int')
calories_burned.dtype

dtype('int64')

In [92]:
calories_burned

array([181, 197,   0, 174, 223, 149, 140,  38,   1,  40, 101, 152, 150,
       113,  49,  86,   6,  99, 143, 125, 129,   6,   9,  10,  72, 150,
       141, 156,  57,  72,  17, 181, 197, 131, 154, 137, 193,  19, 101,
       139, 164, 137,  22,  17,   9, 145, 192, 146, 234, 167,  16,  17,
        32,  35, 220, 116,  23,  44, 131,  86, 194,  60, 121,  76,  93,
        53,  25, 227, 125, 243,  14,  39,  55, 158,   7, 213, 116, 129,
        21,  28,  16, 180, 138, 176,  99,   5,   1,  10,  47,  84,  23,
         4,   0,   0,   0,   0])

**Hours of Sleep**

In [93]:
hours_of_sleep = np.array(hours_of_sleep, dtype = 'int')
hours_of_sleep.dtype

dtype('int64')

**Mood**

`Mood` is a categorical data type. As a name says, categorical data type has two or more categories in it.

Let's check the values of `mood` variable

In [94]:
mood

array(['Neutral', 'Sad', 'Sad', 'Sad', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Sad', 'Sad', 'Sad', 'Happy', 'Sad', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral', 'Neutral',
       'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Neutral',
       'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Neutral', 'Sad', 'Happy', 'Happy', 'Happy',
       'Happy', 'Happy', 'Happy', 'Happy', 'Sad', 'Neutral', 'Neutral',
       'Sad', 'Sad', 'Neutral', 'Neutral', 'Happy', 'Neutral', 'Neutral',
       'Sad', 'Neutral', 'Sad', 'Neutral', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Sad', 'Happy', 'Neutral', 'Happy', 'Neutral', 'Sad', 'Sad', 'Sad',
       'Neutral', 'Neutral', 'Sad', 'Sad', 'Happy', 'Neutral', 'Neutral',
       'Happy'], dtype='<U10')

In [95]:
np.unique(mood)

array(['Happy', 'Neutral', 'Sad'], dtype='<U10')

**Activity Status**

In [96]:
activity_status

array(['Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Active', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Active',
       'Inactive', 'Inactive', 'Inactive', 'Inactive', 'Inactive',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Inactive', 'Inactive',
       'Inactive', 'Inactive', 'Inactive', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Active', 'Active', 'Active', 'Inactive', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Active',
       'Inactive', 'Active', 'Active', 'Active', 'Active', 'Inactive',
       'Inactive', 'Inactive', 'Inacti

### Let's try to get some insights from the data.

#### What's the average step count? 

How can we calculate average? => `.mean()`

In [ ]:
step_count.mean()

User moves an average of 2900 steps a day.

#### On which day the step count was highest?

How will be find it? 

First we find the index of maximum step count and use that index to get the date.

How'll we find the index? =>  

Numpy provides a function `np.argmax()` which returns the index of maximum value element.

Similarly, we have a function `np.argmin()` which returns the index of minimum element.

In [97]:
step_count.argmax()


69

Here 69 is the index of maximum step count element.

In [98]:
date[step_count.argmax()]

'14-12-2017'

Let's check the calorie burnt on the day

In [99]:
calories_burned[step_count.argmax()]

243

Not bad! 243 calories. Let's try to get the number of steps on that day as well

In [100]:
step_count.max()

7422

7k steps!! Sports mode on!

#### Let's try to compare step counts on bad mood days and good mood days

Average step count on Sad mood days

In [101]:
np.mean(step_count[mood == 'Sad'])

2103.0689655172414

In [102]:
np.sort(step_count[mood == 'Sad'])

array([  25,   36,   61,  133,  177,  181,  221,  299,  518,  651,  702,
        753,  799, 1230, 1258, 1580, 1648, 1696, 2822, 3148, 3519, 3721,
       4061, 4340, 4545, 4687, 5461, 6041, 6676])

In [103]:
np.std(step_count[mood == 'Sad'])

2021.2355035376254

Average step count on happy days

In [104]:
np.mean(step_count[mood == 'Happy'])

3392.725

In [105]:
np.sort(step_count[mood == 'Happy'])

array([ 153,  269,  330,  493,  539,  546,  614,  705,  774,  995, 1421,
       1831, 1867, 2203, 2255, 2725, 3149, 3608, 4005, 4033, 4064, 4068,
       4136, 4275, 4421, 4435, 4550, 4683, 4732, 4779, 4880, 5195, 5376,
       5464, 5537, 5934, 5999, 6314, 6930, 7422])

Average step count on sad days - 2103.

Average step count on happy days - 3392

There may be relation between mood and step count

#### Let's try to check inverse. Mood when step count was greater/lesser

Mood when step count > 4000

In [106]:
np.unique(mood[step_count > 4000], return_counts = True)

(array(['Happy', 'Neutral', 'Sad'], dtype='<U10'), array([22,  9,  7]))

Out of 38 days when step count was more than 4000, user was feeling happy on 22 days.

Mood when step count <= 2000

Out of 39 days, when step count was less than 2000, user was feeling sad on 18 days.

#### **There may be a correlation between Mood and step count**

## Reshape in 2D array

#### We saw reshape and flatten. What if i want to convert a matrix to 1D array using `reshape()`
#### Question: What should I pass in `A.reshape()` if I want to use it to convert `A` to 1D vector?

- **(1, 1)?** - **NO** 


- It means we only have a single element


- But **we don't have a single element**

In [107]:
A = np.arange(12).reshape(3,4)

In [113]:
A.reshape(1,12)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])

- So, **(1, 12)?** - **NO** 


- It will **still remain a 2D Matrix with dimensions $1\times12$**

In [114]:
A.reshape(1, 12)

array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])

- **Correct answer is (12)**


- We need a vector of dimension (12,)


- So we need to pass only 1 dimension in `reshape()`

In [115]:
A.reshape(12)

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])

#### So, Be careful while using `reshape()` to convert a Matrix into a 1D vector

#### What will happen if we pass a negative integer in `reshape()`?

In [126]:
A.reshape(6, -1)

array([[ 0,  1],
       [ 2,  3],
       [ 4,  5],
       [ 6,  7],
       [ 8,  9],
       [10, 11]])

#### Surprisingly, it did not give an error

- It is able to **figure out on its own** what should be the **value in-place of negative integer**


- Since **no. of elements in our matrix is 12**


- And **we passed 6 as no. of rows**


- It is **able to figure out** that **no. of columns should be 2**


**Same thing happens with this:**

In [128]:
A.reshape(-1, 3)

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

## Matrix multiplication

#### Question: What will be output of following? 

In [129]:
a = np.arange(5)
b = np.ones(5) * 2

In [130]:
a * b 

array([0., 2., 4., 6., 8.])

Recall that, if a and b are 1D, * operation will perform elementwise multiplication


#### Lets try * with 2D arrays

In [131]:
A = np.arange(12).reshape(3, 4)
A

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [132]:
B = np.arange(12).reshape(3, 4)
B

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [134]:
A * B

array([[  0,   1,   4,   9],
       [ 16,  25,  36,  49],
       [ 64,  81, 100, 121]])

In [136]:
np.dot(A.T, B)

array([[ 80,  92, 104, 116],
       [ 92, 107, 122, 137],
       [104, 122, 140, 158],
       [116, 137, 158, 179]])

**Again did element-wise multiplication**

#### For actual Matrix Multiplication, We have a different method/operator

`np.matmul()`




#### What is the requirement of dimensions of 2 matrices for Matrix Multiplication?

- **Columns of A = Rows of B** (A **Must condition** for Matric Multiplication)


- **If A is $3\times4$, B can be $4\times3$**... or $4\times(Something Else)$

#### So, lets reshape B to $4\times3$ instead

In [137]:
B = B.reshape(4, 3)
B

array([[ 0,  1,  2],
       [ 3,  4,  5],
       [ 6,  7,  8],
       [ 9, 10, 11]])

In [138]:
np.matmul(A, B)

array([[ 42,  48,  54],
       [114, 136, 158],
       [186, 224, 262]])

In [139]:
np.dot(A,B)

array([[ 42,  48,  54],
       [114, 136, 158],
       [186, 224, 262]])

- **We are getting a $3\times3$ matrix as output**

- So, this is doing Matrix Multiplication

#### There's a direct operator as well for Matrix Multiplication
`@`

In [140]:
A @ B

array([[ 42,  48,  54],
       [114, 136, 158],
       [186, 224, 262]])

#### Question: What will be the dimensions of Matrix Multiplication `B @ A`?

- $4\times4$

In [141]:
B @ A

array([[ 20,  23,  26,  29],
       [ 56,  68,  80,  92],
       [ 92, 113, 134, 155],
       [128, 158, 188, 218]])

#### There is another method in np for doing Matrix Multiplication


In [142]:
np.dot(A, B)

array([[ 42,  48,  54],
       [114, 136, 158],
       [186, 224, 262]])

**Other cases of `np.dot()`**
- It performs dot product when both inputs are 1D array
- It performs multiplication when both input are scalers.


In [143]:
a= np.array([1,2,3])
b = np.array([1,1,1])


In [144]:
np.dot(a,b) # 1*1 + 2*1 + 3*1 = 6

6

In [145]:
np.dot(4,5)

20

#### Now, Let's try multiplication of a mix of matrices and vectors



In [146]:
A = np.arange(12).reshape(3, 4)  # A is a 3x4 Matrix 
A

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])

In [147]:
a = np.array([1, 2, 3])  # a although a (3,) can be thought of as row vector
print(a.shape)

(3,)


In [151]:
np.matmul(A, a)

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 4)

**Columns of `A` $\neq$ Rows of `a`**

Lets try revervse

In [153]:
A, a

(array([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]]),
 array([1, 2, 3]))

In [155]:
np.matmul(a, A.reshape(3, -1))

array([32, 38, 44, 50])

YES, **Columns of `a` (3) = Rows of `A` (3)**

In [156]:
a

array([1, 2, 3])

In [158]:
A

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])